packages/gotrue/lib/src/gotrue_client.dart

```diff
@@ -863,30 +863,29 @@ class GoTrueClient {
       final authResponse = AuthResponse.fromJson(response);
 
       if (authResponse.session == null) {
-        final error = AuthException('Invalid session data.');
-        completer.completeError(error, StackTrace.current);
-        throw error;
+        throw AuthException('Invalid session data.');
       }
 
       _saveSession(authResponse.session!);
 
-      completer.complete(authResponse);
       _notifyAllSubscribers(AuthChangeEvent.tokenRefreshed);
+      _refreshTokenCompleter!.complete(authResponse);
+      return authResponse;
     } on SocketException {
       _setTokenRefreshTimer(
         Constants.retryInterval * pow(2, _refreshTokenRetryCount),
-        completer,
         refreshToken: token,
         accessToken: accessToken,
       );
+      return _refreshTokenCompleter!.future;
     } catch (error, stack) {
       if (error is AuthException) {
         if (error.message == 'Invalid Refresh Token: Refresh Token Not Found') {
           await signOut();
         }
       }
-      completer.completeError(error, stack);
       _onAuthStateChangeController.addError(error, stack);
+      rethrow;
```

Comment:
I think developers can listen to the errors with `onStateChanged`, so I'm not too concerned, but thought I would flag it with you!

Reply:
I've added a new commit to catch any errors from `_callRefreshToken` in `_saveSesison` and `_setTokenRefreshTimer`, because these should already be added to the stream and be handled there. Testing this with a test is very complicated, because all these methods are private and I would have to create a cusomt http client to mock some responses.

Reply:
Thank you so much for all the hard work!

Reply:
I wouldn't mind making things public with the [internal](https://api.flutter.dev/flutter/meta/internal-constant.html) annotation like you did with another method or property the other day if that makes testing things easier.

Reply:
I don't think that extra effort is neccessary. I'm very confident the errors are properly catched. `_saveSession` and `_setTokenRefreshTimer` are the only ones calling `_callRefreshToken`, which are not directly called by the user, so I'm pretty sure that's correct now.